File: frontend/nextjs-app/src/lib/services/download-manager-service.ts
import { LocalStorageService, StoragePriority } from './local-storage-service';

/**
 * Download status
 */
export enum DownloadStatus {
  QUEUED = 'queued',
  DOWNLOADING = 'downloading',
  PAUSED = 'paused',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

/**
 * Download item
 */
export interface DownloadItem {
  id: string;
  key: string;
  title: string;
  description?: string;
  priority: StoragePriority;
  estimatedSize: number;
  downloader: () => Promise<Uint8Array>;
  status: DownloadStatus;
  downloadedBytes: number;
  error?: string;
  startedAt?: Date;
  completedAt?: Date;
}

/**
 * Download manager configuration
 */
export interface DownloadManagerConfig {
  maxConcurrentDownloads: number;
  autoRetry: boolean;
  maxRetries: number;
  retryDelay: number;
  wifiOnly: boolean;
}

/**
 * Download progress callback
 */
export type DownloadProgressCallback = (item: DownloadItem) => void;

/**
 * Download manager service for web
 */
export class DownloadManagerService {
  private downloads: Map<string, DownloadItem> = new Map();
  private queue: string[] = [];
  private activeDownloads: Set<string> = new Set();
  private retryCount: Map<string, number> = new Map();
  private listeners: Set<DownloadProgressCallback> = new Set();

  private config: DownloadManagerConfig = {
    maxConcurrentDownloads: 3,
    autoRetry: true,
    maxRetries: 3,
    retryDelay: 5000,
    wifiOnly: false,
  };

  constructor(config?: Partial<DownloadManagerConfig>) {
    if (config) {
      this.config = { ...this.config, ...config };
    }
  }

  /**
   * Subscribe to download updates
   */
  subscribe(callback: DownloadProgressCallback): () => void {
    this.listeners.add(callback);
    return () => this.listeners.delete(callback);
  }

  /**
   * Queue a download
   */
  async queueDownload(params: {
    key: string;
    title: string;
    description?: string;
    priority: StoragePriority;
    estimatedSize: number;
    downloader: () => Promise<Uint8Array>;
  }): Promise<string> {
    const id = Date.now().toString() + Math.random().toString(36).substr(2, 9);

    const item: DownloadItem = {
      id,
      key: params.key,
      title: params.title,
      description: params.description,
      priority: params.priority,
      estimatedSize: params.estimatedSize,
      downloader: params.downloader,
      status: DownloadStatus.QUEUED,
      downloadedBytes: 0,
    };

    this.downloads.set(id, item);
    this.queue.push(id);

    this.notifyListeners();
    this.processQueue();

    return id;
  }

  /**
   * Start/resume a download
   */
  async startDownload(id: string): Promise<void> {
    const item = this.downloads.get(id);
    if (!item) return;

    if (item.status === DownloadStatus.PAUSED) {
      item.status = DownloadStatus.QUEUED;
      if (!this.queue.includes(id)) {
        this.queue.push(id);
      }
      this.notifyListeners();
      this.processQueue();
    }
  }

  /**
   * Pause a download
   */
  async pauseDownload(id: string): Promise<void> {
    const item = this.downloads.get(id);
    if (!item) return;

    if (
      item.status === DownloadStatus.DOWNLOADING ||
      item.status === DownloadStatus.QUEUED
    ) {
      item.status = DownloadStatus.PAUSED;
      this.queue = this.queue.filter((qid) => qid !== id);
      this.activeDownloads.delete(id);
      this.notifyListeners();
    }
  }

  /**
   * Cancel a download
   */
  async cancelDownload(id: string): Promise<void> {
    const item = this.downloads.get(id);
    if (!item) return;

    item.status = DownloadStatus.CANCELLED;
    this.queue = this.queue.filter((qid) => qid !== id);
    this.activeDownloads.delete(id);
    this.downloads.delete(id);
    this.retryCount.delete(id);

    this.notifyListeners();
  }

  /**
   * Retry a failed download
   */
  async retryDownload(id: string): Promise<void> {
    const item = this.downloads.get(id);
    if (!item || item.status !== DownloadStatus.FAILED) return;

    item.status = DownloadStatus.QUEUED;
    item.error = undefined;
    item.downloadedBytes = 0;
    this.retryCount.set(id, 0);

    if (!this.queue.includes(id)) {
      this.queue.push(id);
    }

    this.notifyListeners();
    this.processQueue();
  }

  /**
   * Clear completed downloads
   */
  async clearCompleted(): Promise<void> {
    const completedIds: string[] = [];

    this.downloads.forEach((item, id) => {
      if (item.status === DownloadStatus.COMPLETED) {
        completedIds.push(id);
      }
    });

    completedIds.forEach((id) => this.downloads.delete(id));
    this.notifyListeners();
  }

  /**
   * Get all downloads
   */
  getDownloads(): DownloadItem[] {
    return Array.from(this.downloads.values());
  }

  /**
   * Get download by ID
   */
  getDownload(id: string): DownloadItem | undefined {
    return this.downloads.get(id);
  }

  /**
   * Get active downloads
   */
  getActiveDownloads(): DownloadItem[] {
    return this.getDownloads().filter(
      (d) =>
        d.status === DownloadStatus.DOWNLOADING ||
        d.status === DownloadStatus.QUEUED
    );
  }

  /**
   * Get completed downloads
   */
  getCompletedDownloads(): DownloadItem[] {
    return this.getDownloads().filter(
      (d) => d.status === DownloadStatus.COMPLETED
    );
  }

  /**
   * Get failed downloads
   */
  getFailedDownloads(): DownloadItem[] {
    return this.getDownloads().filter((d) => d.status === DownloadStatus.FAILED);
  }

  /**
   * Get total download size
   */
  getTotalSize(): number {
    return this.getDownloads().reduce(
      (sum, item) => sum + item.estimatedSize,
      0
    );
  }

  /**
   * Get downloaded size
   */
  getDownloadedSize(): number {
    return this.getDownloads().reduce(
      (sum, item) => sum + item.downloadedBytes,
      0
    );
  }

  /**
   * Get overall progress
   */
  getOverallProgress(): number {
    const total = this.getTotalSize();
    if (total === 0) return 0;
    return Math.min(1, this.getDownloadedSize() / total);
  }

  /**
   * Process download queue
   */
  private async processQueue(): Promise<void> {
    // Check if we can start more downloads
    while (
      this.activeDownloads.size < this.config.maxConcurrentDownloads &&
      this.queue.length > 0
    ) {
      // Sort queue by priority
      this.queue.sort((a, b) => {
        const itemA = this.downloads.get(a);
        const itemB = this.downloads.get(b);
        if (!itemA || !itemB) return 0;
        return itemA.priority - itemB.priority;
      });

      const id = this.queue.shift();
      if (!id) continue;

      const item = this.downloads.get(id);
      if (!item || item.status !== DownloadStatus.QUEUED) {
        continue;
      }

      this.activeDownloads.add(id);
      this.downloadItem(id);
    }
  }

  /**
   * Download an item
   */
  private async downloadItem(id: string): Promise<void> {
    const item = this.downloads.get(id);
    if (!item) return;

    try {
      item.status = DownloadStatus.DOWNLOADING;
      item.startedAt = new Date();
      this.notifyListeners();

      // Download data
      const data = await item.downloader();

      // Store in local storage
      await LocalStorageService.store(
        item.key,
        Array.from(data),
        item.priority
      );

      item.status = DownloadStatus.COMPLETED;
      item.completedAt = new Date();
      item.downloadedBytes = item.estimatedSize;
      this.retryCount.delete(id);

      console.log('Download completed:', item.title);
    } catch (error) {
      console.error('Download failed:', item.title, error);

      const retries = this.retryCount.get(id) || 0;

      if (this.config.autoRetry && retries < this.config.maxRetries) {
        this.retryCount.set(id, retries + 1);
        item.status = DownloadStatus.QUEUED;

        // Add back to queue after delay
        setTimeout(() => {
          if (!this.queue.includes(id)) {
            this.queue.push(id);
          }
          this.processQueue();
        }, this.config.retryDelay);

        console.log(
          `Retrying download: ${item.title} (${retries + 1}/${this.config.maxRetries})`
        );
      } else {
        item.status = DownloadStatus.FAILED;
        item.error = error instanceof Error ? error.message : String(error);
      }
    } finally {
      this.activeDownloads.delete(id);
      this.notifyListeners();
      this.processQueue();
    }
  }

  /**
   * Notify listeners
   */
  private notifyListeners(): void {
    this.downloads.forEach((item) => {
      this.listeners.forEach((callback) => callback(item));
    });
  }
}

// Singleton instance
let downloadManagerInstance: DownloadManagerService | null = null;

export function getDownloadManager(): DownloadManagerService {
  if (!downloadManagerInstance) {
    downloadManagerInstance = new DownloadManagerService();
  }
  return downloadManagerInstance;
}

File: frontend/mobile/lib/core/services/download_manager_service.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'local_storage_service.dart';

/// Download status
enum DownloadStatus {
  queued,
  downloading,
  paused,
  completed,
  failed,
  cancelled,
}

/// Download item
class DownloadItem {
  final String id;
  final String key;
  final String title;
  final String? description;
  final StoragePriority priority;
  final int estimatedSize;
  final Future<List<int>> Function() downloader;
  
  DownloadStatus status;
  int downloadedBytes;
  String? error;
  DateTime? startedAt;
  DateTime? completedAt;
  List<DownloadChunk>? chunks;
  double? downloadSpeed; // bytes per second
  double? remainingTime; // seconds

  DownloadItem({
    required this.id,
    required this.key,
    required this.title,
    this.description,
    required this.priority,
    required this.estimatedSize,
    required this.downloader,
    this.status = DownloadStatus.queued,
    this.downloadedBytes = 0,
    this.error,
    this.startedAt,
    this.completedAt,
    this.chunks,
    this.downloadSpeed,
    this.remainingTime,
  });

  double get progress {
    if (estimatedSize == 0) return 0;
    return (downloadedBytes / estimatedSize).clamp(0.0, 1.0);
  }

  bool get isActive => status == DownloadStatus.downloading || status == DownloadStatus.queued;
  bool get isCompleted => status == DownloadStatus.completed;
  bool get isFailed => status == DownloadStatus.failed;
  bool get isPaused => status == DownloadStatus.paused;

  DownloadItem copyWith({
    DownloadStatus? status,
    int? downloadedBytes,
    String? error,
    DateTime? startedAt,
    DateTime? completedAt,
    List<DownloadChunk>? chunks,
    double? downloadSpeed,
    double? remainingTime,
  }) {
    return DownloadItem(
      id: id,
      key: key,
      title: title,
      description: description,
      priority: priority,
      estimatedSize: estimatedSize,
      downloader: downloader,
      status: status ?? this.status,
      downloadedBytes: downloadedBytes ?? this.downloadedBytes,
      error: error ?? this.error,
      startedAt: startedAt ?? this.startedAt,
      completedAt: completedAt ?? this.completedAt,
      chunks: chunks ?? this.chunks,
      downloadSpeed: downloadSpeed ?? this.downloadSpeed,
      remainingTime: remainingTime ?? this.remainingTime,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'key': key,
        'title': title,
        'description': description,
        'priority': priority.index,
        'estimatedSize': estimatedSize,
        'status': status.index,
        'downloadedBytes': downloadedBytes,
        'error': error,
        'startedAt': startedAt?.toIso8601String(),
        'completedAt': completedAt?.toIso8601String(),
        'downloadSpeed': downloadSpeed,
        'remainingTime': remainingTime,
      };
}

/// Download chunk for progressive loading
class DownloadChunk {
  final int index;
  final int start;
  final int end;
  bool downloaded;

  DownloadChunk({
    required this.index,
    required this.start,
    required this.end,
    this.downloaded = false,
  });
}

/// Download manager configuration
class DownloadManagerConfig {
  final int maxConcurrentDownloads;
  final bool autoRetry;
  final int maxRetries;
  final Duration retryDelay;
  final bool wifiOnly;
  final int chunkSize; // Size of each chunk for progressive download
  final bool enableProgressiveDownload;

  const DownloadManagerConfig({
    this.maxConcurrentDownloads = 3,
    this.autoRetry = true,
    this.maxRetries = 3,
    this.retryDelay = const Duration(seconds: 5),
    this.wifiOnly = false,
    this.chunkSize = 1024 * 1024, // 1MB chunks
    this.enableProgressiveDownload = true,
  });
}

/// Download manager service
class DownloadManagerService {
  final LocalStorageService _storageService;
  final DownloadManagerConfig config;
  
  final Map<String, DownloadItem> _downloads = {};
  final List<String> _queue = [];
  final Set<String> _activeDownloads = {};
  final Map<String, int> _retryCount = {};
  
  final StreamController<List<DownloadItem>> _downloadsController = 
      StreamController<List<DownloadItem>>.broadcast();

  DownloadManagerService(
    this._storageService, {
    this.config = const DownloadManagerConfig(),
  });

  /// Stream of all downloads
  Stream<List<DownloadItem>> get downloadsStream => _downloadsController.stream;

  /// Get all downloads
  List<DownloadItem> get downloads => _downloads.values.toList();

  /// Get active downloads
  List<DownloadItem> get activeDownloads => 
      _downloads.values.where((d) => d.isActive).toList();

  /// Get completed downloads
  List<DownloadItem> get completedDownloads => 
      _downloads.values.where((d) => d.isCompleted).toList();

  /// Get failed downloads
  List<DownloadItem> get failedDownloads => 
      _downloads.values.where((d) => d.isFailed).toList();

  /// Queue a download
  Future<String> queueDownload({
    required String key,
    required String title,
    String? description,
    required StoragePriority priority,
    required int estimatedSize,
    required Future<List<int>> Function() downloader,
  }) async {
    final id = DateTime.now().millisecondsSinceEpoch.toString();
    
    final item = DownloadItem(
      id: id,
      key: key,
      title: title,
      description: description,
      priority: priority,
      estimatedSize: estimatedSize,
      downloader: downloader,
    );

    _downloads[id] = item;
    _queue.add(id);
    
    _notifyListeners();
    _processQueue();

    return id;
  }

  /// Start/resume a download
  Future<void> startDownload(String id) async {
    final item = _downloads[id];
    if (item == null) return;

    if (item.status == DownloadStatus.paused) {
      item.status = DownloadStatus.queued;
      if (!_queue.contains(id)) {
        _queue.add(id);
      }
      _notifyListeners();
      _processQueue();
    }
  }

  /// Pause a download
  Future<void> pauseDownload(String id) async {
    final item = _downloads[id];
    if (item == null) return;

    if (item.status == DownloadStatus.downloading || 
        item.status == DownloadStatus.queued) {
      item.status = DownloadStatus.paused;
      _queue.remove(id);
      _activeDownloads.remove(id);
      _notifyListeners();
    }
  }

  /// Cancel a download
  Future<void> cancelDownload(String id) async {
    final item = _downloads[id];
    if (item == null) return;

    item.status = DownloadStatus.cancelled;
    _queue.remove(id);
    _activeDownloads.remove(id);
    _downloads.remove(id);
    _retryCount.remove(id);
    
    _notifyListeners();
  }

  /// Retry a failed download
  Future<void> retryDownload(String id) async {
    final item = _downloads[id];
    if (item == null || !item.isFailed) return;

    item.status = DownloadStatus.queued;
    item.error = null;
    item.downloadedBytes = 0;
    _retryCount[id] = 0;
    
    if (!_queue.contains(id)) {
      _queue.add(id);
    }
    
    _notifyListeners();
    _processQueue();
  }

  /// Clear completed downloads
  Future<void> clearCompleted() async {
    final completedIds = _downloads.entries
        .where((e) => e.value.isCompleted)
        .map((e) => e.key)
        .toList();

    for (var id in completedIds) {
      _downloads.remove(id);
    }

    _notifyListeners();
  }

  /// Get download by ID
  DownloadItem? getDownload(String id) => _downloads[id];

  /// Get total download size
  int getTotalSize() {
    return _downloads.values.fold(0, (sum, item) => sum + item.estimatedSize);
  }

  /// Get downloaded size
  int getDownloadedSize() {
    return _downloads.values.fold(0, (sum, item) => sum + item.downloadedBytes);
  }

  /// Get overall progress
  double getOverallProgress() {
    final total = getTotalSize();
    if (total == 0) return 0;
    return (getDownloadedSize() / total).clamp(0.0, 1.0);
  }

  /// Estimate required space for pending downloads
  int getRequiredSpace() {
    return _downloads.values
        .where((d) =>
            d.status == DownloadStatus.queued ||
            d.status == DownloadStatus.downloading ||
            d.status == DownloadStatus.paused)
        .fold(0, (sum, item) => sum + (item.estimatedSize - item.downloadedBytes));
  }

  /// Check if there's enough space for downloads
  Future<bool> hasEnoughSpace() async {
    final required = getRequiredSpace();
    final stats = await _storageService.getStats();
    final available = stats.totalSize - stats.usedSpace;
    return available >= required;
  }

  /// Get space availability info
  Future<SpaceInfo> getSpaceInfo() async {
    final required = getRequiredSpace();
    final stats = await _storageService.getStats();
    final available = stats.totalSize - stats.usedSpace;
    final hasEnough = available >= required;
    final deficit = hasEnough ? 0 : required - available;

    return SpaceInfo(
      required: required,
      available: available,
      hasEnough: hasEnough,
      deficit: deficit,
    );
  }

  /// Get estimated completion time for all downloads
  double getEstimatedCompletionTime() {
    final active = activeDownloads;
    if (active.isEmpty) return 0;

    final totalRemaining = active.fold<double>(
      0,
      (sum, item) => sum + (item.remainingTime ?? 0),
    );

    return totalRemaining / active.length;
  }

  /// Process download queue
  Future<void> _processQueue() async {
    // Check if we can start more downloads
    while (_activeDownloads.length < config.maxConcurrentDownloads && 
           _queue.isNotEmpty) {
      // Sort queue by priority
      _queue.sort((a, b) {
        final itemA = _downloads[a];
        final itemB = _downloads[b];
        if (itemA == null || itemB == null) return 0;
        return itemA.priority.index.compareTo(itemB.priority.index);
      });

      final id = _queue.removeAt(0);
      final item = _downloads[id];
      
      if (item == null || item.status != DownloadStatus.queued) {
        continue;
      }

      _activeDownloads.add(id);
      _downloadItem(id);
    }
  }

  /// Download an item
  Future<void> _downloadItem(String id) async {
    final item = _downloads[id];
    if (item == null) return;

    try {
      item.status = DownloadStatus.downloading;
      item.startedAt = DateTime.now();
      _notifyListeners();

      final startTime = DateTime.now();

      // Download with progress tracking
      await _downloadWithProgress(item, (progress) {
        final elapsed = DateTime.now().difference(startTime).inSeconds;
        if (elapsed > 0) {
          final speed = progress / elapsed.toDouble();
          final remaining = (item.estimatedSize - progress) / speed;

          item.downloadedBytes = progress;
          item.downloadSpeed = speed;
          item.remainingTime = remaining;
          _notifyListeners();
        }
      });

      item.status = DownloadStatus.completed;
      item.completedAt = DateTime.now();
      item.downloadedBytes = item.estimatedSize;
      item.downloadSpeed = null;
      item.remainingTime = null;
      _retryCount.remove(id);
      
      debugPrint('Download completed: ${item.title}');
    } catch (e) {
      debugPrint('Download failed: ${item.title} - $e');
      
      final retries = _retryCount[id] ?? 0;
      
      if (config.autoRetry && retries < config.maxRetries) {
        _retryCount[id] = retries + 1;
        item.status = DownloadStatus.queued;
        
        // Add back to queue after delay
        await Future.delayed(config.retryDelay);
        if (!_queue.contains(id)) {
          _queue.add(id);
        }
        
        debugPrint('Retrying download: ${item.title} (${retries + 1}/${config.maxRetries})');
      } else {
        item.status = DownloadStatus.failed;
        item.error = e.toString();
      }
    } finally {
      _activeDownloads.remove(id);
      _notifyListeners();
      _processQueue();
    }
  }

  /// Download with progress tracking
  Future<void> _downloadWithProgress(
    DownloadItem item,
    void Function(int bytes) onProgress,
  ) async {
    if (config.enableProgressiveDownload && item.estimatedSize > config.chunkSize) {
      await _downloadProgressive(item, onProgress);
    } else {
      // Simple download for small files
      await _storageService.downloadContent(
        item.key,
        item.downloader,
        priority: item.priority,
        onProgress: (received, total) {
          onProgress(received);
        },
      );
    }
  }

  /// Progressive download in chunks
  Future<void> _downloadProgressive(
    DownloadItem item,
    void Function(int bytes) onProgress,
  ) async {
    // Initialize chunks if not already done
    if (item.chunks == null) {
      final numChunks = (item.estimatedSize / config.chunkSize).ceil();
      item.chunks = List.generate(
        numChunks,
        (i) => DownloadChunk(
          index: i,
          start: i * config.chunkSize,
          end: ((i + 1) * config.chunkSize).clamp(0, item.estimatedSize),
        ),
      );
    }

    // Download content with chunk tracking
    await _storageService.downloadContent(
      item.key,
      item.downloader,
      priority: item.priority,
      onProgress: (received, total) {
        // Update chunk status
        if (item.chunks != null) {
          for (final chunk in item.chunks!) {
            if (received >= chunk.end) {
              chunk.downloaded = true;
            }
          }
        }
        onProgress(received);
      },
    );
  }

  /// Notify listeners
  void _notifyListeners() {
    _downloadsController.add(downloads);
  }

  /// Dispose
  void dispose() {
    _downloadsController.close();
  }
}

/// Space availability information
class SpaceInfo {
  final int required;
  final int available;
  final bool hasEnough;
  final int deficit;

  const SpaceInfo({
    required this.required,
    required this.available,
    required this.hasEnough,
    required this.deficit,
  });

  Map<String, dynamic> toJson() => {
        'required': required,
        'available': available,
        'hasEnough': hasEnough,
        'deficit': deficit,
      };
}
